File: easy/ts/408.ts
// https://leetcode.com/problems/valid-word-abbreviation/
// Time: O(n + m), n===word.length, m===abbr.length
// Space: O(1)
function validWordAbbreviation(word: string, abbr: string): boolean {
  let i = 0;
  let currAbbrIndex = 0;
  while (i < word.length && currAbbrIndex < abbr.length) {
    if (word[i] === abbr[currAbbrIndex]) {
      i++;
      currAbbrIndex++;
    } else if (!isNaN(Number(abbr[currAbbrIndex]))) {
      if (Number(abbr[currAbbrIndex]) === 0) return false;
      if (abbr[currAbbrIndex + 1] && !isNaN(Number(abbr[currAbbrIndex + 1]))) {
        if (abbr[currAbbrIndex + 2] && !isNaN(Number(abbr[currAbbrIndex + 2])))
          return false;
        let numberToSkip = Number(
          abbr[currAbbrIndex] + abbr[currAbbrIndex + 1]
        );
        if (i + numberToSkip > word.length) return false;
        i += numberToSkip;
        currAbbrIndex += 2;
      } else {
        let numberToSkip = Number(abbr[currAbbrIndex]);
        if (i + numberToSkip > word.length) return false;
        i += numberToSkip;
        currAbbrIndex += 1;
      }
    } else {
      return false;
    }
  }
  return i === word.length && currAbbrIndex === abbr.length;
}

File: easy/python/408.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        i = 0
        abbr_index = 0
        while i < len(word) and abbr_index < len(abbr):
            if word[i] == abbr[abbr_index]:
                i += 1
                abbr_index += 1
            elif abbr[abbr_index].isdigit():
                if abbr[abbr_index + 1] in abbr and abbr[abbr_index + 1].isdigit():
                    integer_num = int(abbr[abbr_index] + abbr[abbr_index + 1])
                    if i + integer_num > len(word) or (
                        abbr[abbr_index + 2] in abbr and abbr[abbr_index + 2].isdigit()
                    ):
                        return False
                    i += integer_num
                    abbr_index += 2
                else:
                    integer_num = int(abbr[abbr_index])
                    if i + integer_num > len(word):
                        return False
                    i += integer_num
                    abbr_index += 1
            else:
                return False
        if i == len(word) and abbr_index == len(abbr):
            return True
        return False
